dask/dataframe/io/csv.py

```diff
@@ -183,6 +191,36 @@ def text_blocks_to_pandas(reader, block_lists, header, head, kwargs,
     A dask.dataframe or list of delayed values
     """
     dtypes = head.dtypes.to_dict()
+    # dtypes contains only instances of CategoricalDtype, which causes issues
+    # in coerce_dtypes for non-uniform categories accross partitions.
+    # We will modify `dtype` (which is inferred) to
+    # 1. contain instances of CategoricalDtypes for user-provided types
+    # 2. contain 'category' for data inferred types
+    categoricals = head.select_dtypes(include=['category']).columns
+
+    if isinstance(specified_dtypes, collections.Mapping):
```

Comment:
Since we can only have known categoricals if `HAS_CDT` is true, I'd rather switch this logic to all be under a `if _HAS_CDT` block, and dispatch for mapping/non-mapping from there. Then you don't have to check `_HAS_CDT` in the loop below either.